packages/gatsby/src/schema/types/NodeInterface.js

```diff
@@ -0,0 +1,76 @@
+const getOrCreateNodeInterface = schemaComposer => {
+  // TODO: why is `mediaType` on Internal? Applies only to File!?
+  // Is `fieldOwners` actually set anywhere? And is it supposed to be an array,
+  // as the Joi schema claims, or an object, which is the default in `redux/actions.js`
+  const InternalTC = schemaComposer.getOrCreateTC(`Internal`, tc => {
+    tc.addFields({
+      content: `String`,
+      contentDigest: `String`,
+      description: `String`,
+      fieldOwners: [`String`],
+      ignoreType: `Boolean`,
+      mediaType: `String`,
+      owner: `String`,
+      type: `String!`,
+    })
+  })
+  InternalTC.getITC()
+
+  const NodeInterfaceTC = schemaComposer.getOrCreateIFTC(`Node`, tc => {
+    tc.setDescription(`Node Interface`)
+    tc.addFields({
+      id: `String!`,
+      parent: {
+        type: `Node`,
+        resolve: async (source, args, context, info) =>
+          context.nodeModel.getNode(source.parent),
+      },
+      children: {
+        type: `[Node]!`,
+        resolve: async (source, args, context, info) =>
+          context.nodeModel.getNodes(source.children),
+      },
+      internal: `Internal`,
+    })
+  })
+  NodeInterfaceTC.getITC()
+
+  return NodeInterfaceTC
+}
+
+const addNodeInterface = ({ schemaComposer, typeComposer }) => {
+  const NodeInterfaceTC = getOrCreateNodeInterface(schemaComposer)
+  typeComposer.addInterface(NodeInterfaceTC)
+  NodeInterfaceTC.addTypeResolver(
+    typeComposer,
+    node => node.internal.type === typeComposer.getTypeName()
+  )
+  addNodeInterfaceFields({ schemaComposer, typeComposer })
+}
+
+const addNodeInterfaceFields = ({ schemaComposer, typeComposer }) => {
+  const NodeInterfaceTC = getOrCreateNodeInterface(schemaComposer)
+  typeComposer.addFields(NodeInterfaceTC.getFields())
+  NodeInterfaceTC.addTypeResolver(
```

Comment:
Although we maybe also want to add `typeComposer.isTypeOf` to the types that implement the Node interface?? Might be beneficial in the future when we allow custom interfaces so the correct types are resolved automatically?

Reply:
I'd rather use typeResolver than isTypeOf - isTypeOf is called for every possible type of the interface whenever an object is checked, thus it will be a O(N) checks for N types.